### PR TITLE
[Core] Improve JsonOutputProcess

### PR DESCRIPTION
**Pull request by loumalouomega**
**📝 Description**

Improve JsonOutputProcess. Avoiding write when not required (not elements, not GP). Reduce code duplication. Improve dump json.

**🆕 Changelog**

- [Improve JsonOutputProcess](https://github.com/KratosMultiphysics/Kratos/commit/5a9591e37b38ce44d2247ee62dbbb58bb164963b)
- Avoiding write when not required (not elements, not GP)
- Reduce code duplication
- [Improve dump json](https://github.com/KratosMultiphysics/Kratos/pull/10562/commits/71cd223858de44760d3a95243a8954164f388010)

### CODE REVIEW

File: kratos/python_scripts/json_output_process.py [hunk 1]
```diff
@@ -100,129 +100,130 @@ def ExecuteBeforeSolutionLoop(self):
...
         data = {}
         data["TIME"] = []
-        count = 0
-        for node in self.sub_model_part.Nodes:
-            compute = self.__check_flag(node)
-
-            if compute:
-                node_identifier = "NODE_" + self.__get_node_identifier(node)
+        # Nodal values
+        if len(self.output_variables) > 0:
```
Comment (philbucher) [inline kratos/python_scripts/json_output_process.py]:
```suggestion
        if self.output_variables:
```

File: kratos/python_scripts/json_output_process.py [hunk 2]
```diff
@@ -100,129 +100,130 @@ def ExecuteBeforeSolutionLoop(self):
...
-                            for gp in range(gauss_point_number):
-                                data["ELEMENT_" + str(elem.Id)][variable_name][str(gp)] = []
-                        else:
-                            if count == 0:
-                                data["RESULTANT"][variable_name] = {}
-                                for gp in range(gauss_point_number):
-                                    data["RESULTANT"][variable_name][str(gp)] = []
-                    elif variable_type == "Array":
-                        if KratosMultiphysics.KratosGlobals.GetVariableType(variable_name + "_X") == "Double":
+        if len(self.gauss_points_output_variables) > 0:
```
Comment (philbucher) [inline kratos/python_scripts/json_output_process.py]:
```suggestion
        if self.gauss_points_output_variables:
```

File: kratos/python_scripts/json_utilities.py [hunk 3]
```diff
@@ -8,5 +7,4 @@ def read_external_json(file_name):
 
 def write_external_json(file_name, data):
   with open(file_name, 'w') as outfile:
-    json.dump(data, outfile)
-
+    json.dump(data, outfile, indent=4)
```
Comment (philbucher) [inline kratos/python_scripts/json_utilities.py]:
```suggestion
    json.dump(data, outfile, indent=2)
```

4 is quite verbose

File: kratos/python_scripts/json_output_process.py [hunk 4]
```diff
@@ -243,147 +244,147 @@ def ExecuteFinalizeSolutionStep(self):
...
 
-                if compute:
-                    for i in range(self.params["gauss_points_output_variables"].size()):
-                        out = self.params["gauss_points_output_variables"][i]
-                        variable_name = out.GetString()
-                        variable = KratosMultiphysics.KratosGlobals.GetVariable(variable_name)
-                        variable_type = KratosMultiphysics.KratosGlobals.GetVariableType(variable_name)
-                        value = elem.CalculateOnIntegrationPoints(variable, self.sub_model_part.ProcessInfo)
+                    if compute:
+                        for i in range(len(self.gauss_points_output_variables)):
```
Comment (philbucher) [inline kratos/python_scripts/json_output_process.py]: enumerate

File: kratos/python_scripts/json_output_process.py [hunk 5]
```diff
@@ -243,147 +244,147 @@ def ExecuteFinalizeSolutionStep(self):
...
-                        # TODO: Add pending classes
-                    count += 1
+                            # TODO: Add pending classes
+                        count += 1
 
             count = 0
             # Gauss points values
-            for elem in self.sub_model_part.Elements:
-                compute = self.__check_flag(elem)
+            if len(self.gauss_points_output_variables) > 0:
```
Comment (philbucher) [inline kratos/python_scripts/json_output_process.py]:
```suggestion
            if self.gauss_points_output_variables:
```

File: kratos/python_scripts/json_output_process.py [hunk 6]
```diff
@@ -243,147 +244,147 @@ def ExecuteFinalizeSolutionStep(self):
...
-                        out = self.params["gauss_points_output_variables"][i]
-                        variable_name = out.GetString()
-                        variable = KratosMultiphysics.KratosGlobals.GetVariable(variable_name)
-                        variable_type = KratosMultiphysics.KratosGlobals.GetVariableType(variable_name)
-                        value = elem.CalculateOnIntegrationPoints(variable, self.sub_model_part.ProcessInfo)
+                    if compute:
+                        for i in range(len(self.gauss_points_output_variables)):
+                            variable = self.gauss_points_output_variables[i]
+                            variable_name = variable.Name()
+                            variable_type = KratosMultiphysics.KratosGlobals.GetVariableType(variable_name)
```
Comment (philbucher) [inline kratos/python_scripts/json_output_process.py:319]: this is slow, could it be done outside the loop?

**Review by philbucher — COMMENTED** [review]
some minor comments

File: kratos/python_scripts/json_output_process.py (hunk 6 above)
Comment (loumalouomega) [reply kratos/python_scripts/json_output_process.py:319]: Maybe

Comment (loumalouomega) [reply kratos/python_scripts/json_output_process.py:319]: Thinking thinking

Comment (philbucher) [reply kratos/python_scripts/json_output_process.py:319]: please dont refactor the entire class, it can also be left as is :D

**Review by philbucher — APPROVED** [review]

File: kratos/python_scripts/json_output_process.py (hunk 6 above)
Comment (loumalouomega) [reply kratos/python_scripts/json_output_process.py:319]: I have an idea, for a next PR

**Pull request merged**